SHM Transport documentation update <master> [8110] (#104)
eProsima/Fast-RTPS-docs: docs/advanced.rst

*eProsima Fast RTPS* implements an architecture of pluggable transports.

> Current version implements five transports: UDPv4, UDPv6, TCPv4, TCPv6 and SHM.

Reviewer: ```suggestion
Current version implements five transports: UDPv4, UDPv6, TCPv4, TCPv6 and SHM (shared memory).
```

> By default, when a :class:`Participant` is created, two built-in transports are
> configured SHM and UDPv4. SHM transport will be used for all communications between
> participants in the same machine, and UDPv4 for external communications.

Reviewer 2: ```suggestion
By default, when a :class:`Participant` is created, two built-in transports are configured:

* SHM transport will be used for all communications between participants in the same machine.
* UDPv4 will be used for inter machine communications.
```

> configured SHM and UDPv4. SHM transport will be used for all communications between
> participants in the same machine, and UDPv4 for external communications.

Reviewer: ```suggestion
configured: SHM and UDPv4. SHM transport is be used for all communications between
participants in the same machine, while UDPv4 is used for external communications.
```

Author: "is be used?" I guess you mean "is used"

All Transport configuration options can be found in the section :ref:`transportdescriptors`.

.. _comm-transports-shm:

Shared memory Transport (SHM)
=============================

The shared memory transport enables fast communications between entities running in the same processing unit/machine,
relying on the shared memory mechanisms provided by the host operating system.

> SHM transport provides better performance than other transports like UDP / TCP, even when these transports use loopback
> interface. This is mainly due to the following reasons:

Reviewer 2: ```suggestion
SHM transport provides better performance than other transports like UDP / TCP, even when these transports use loopback
interface.
This is mainly due to the following reasons:
```

 * Large message support: Network protocols need to fragment data in order to comply with the specific protocol and
   network stacks requirements.
   SHM transport allows the copy of full messages where the only size limit is the machine's memory capacity.

 * Reduce the number of memory copies: When sending the same message to different endpoints, SHM transport can
   directly share the same memory buffer with all the destination endpoints.
   Other protocols require to perform one copy of the message per endpoint.

 * Less operating system overhead: Once initial setup is completed, shared memory transfers require much less system
   calls than the other protocols.
   Therefore there is a performance/time consume gain by using SHM.

When two participants on the same machine have SHM transport enabled, all communications between them are automatically
performed by SHM transport only.
The rest of the enabled transports are not used between those two participants.

To enable SHM transport in a Participant, you need to add the SharedMemTransportDescriptor to the
``rtps.userTransports`` attribute (C++ code) or define a transport_descriptor of type SHM in the

> XML file (see below examples).

Reviewer 2: This will not be the case anymore, as SHM will be active by default. We may mention how to proceed if the user wants to change the default parameters of the transport, though.

Author: you're so right

+--------------------------------------------------+
| **XML**                                          |
+--------------------------------------------------+
| .. literalinclude:: ../code/XMLTester.xml        |
|    :language: xml                                |
|    :start-after: <!-->CONF-SHM-TRANSPORT-SETTING |
|    :end-before: <!--><-->                        |
+--------------------------------------------------+

SHM configuration parameters:

 * ``segment_size``: The size of the shared memory segment in bytes.
   A shared memory segment is created by each participant.
   Participant's writers copy their messages into the segment and send a message reference to the destination readers.

 * ``port_queue_capacity``: Each participant with SHM transport enabled listens on a queue (port) for incoming SHM
   message references.
   This parameter specifies the queue size (in messages).

 * ``healthy_check_timeout_ms``: With SHM, Readers and writers use a queue to exchange messages (called Port).
   If one of the processes involved crashes while using the port, the structure can be left inoperative.
   For this reason, every time a port is opened, a healthy check is performed.

>    If the attached processes doesn't respond in ``healthy_check_timeout_ms`` milliseconds, the port is destroyed and

Reviewer 2: ```suggestion
   If the attached process doesn't respond in ``healthy_check_timeout_ms`` milliseconds, the port is destroyed and
```

Author: There can be may processes attached to a multicast port.

Author: Although it should really be "If the attached processes don't"
